### PR TITLE
feat: assign role

### PR DESCRIPTION
**Pull request by davidolanrewaju**
## What does this PR do?

This PR implements an endpoint for assigning roles to members of an organisation:

* Added `PUT /api/v1/organization/members/{memberId}/role` Endpoint: Assigns roles to members of an organisation .
* Authentication: Ensures only authenticated users can access this endpoint.
* Validation: Implements input validation for the permission list obtained from the client.
* Error Handling: Returns specific HTTP status codes and messages based on various scenarios.

## Checklist:

* [x] Implemented `PUT /api/v1/organization/members/{memberId}/role` endpoint in `organisation.controller.ts`.
* [x] Tested endpoint functionality to ensure it correctly handles various scenarios.
* [x] Added appropriate HTTP status codes and response messages for edge cases.
* [x] Added Swagger documentation for the new endpoint.
* [x] Added unit tests for the endpoint
* [x] Updated API documentation to reflect the new endpoint.

## Response Example

**Request Body**

```json
{
  "role": "string"
}
```

**Success**

```json
{
  "message": "string",
  "status_code": "int",
  "data": {
    "user": {},
    "org": {},
    "role": {}
  }
}
```

**Error (Memeber Does Not Exists)**
```json
{
    "status_code": "int",
    "message": "string"
}
```

**Error (Role Already Assigned)**
```json
{
    "status_code": "int",
    "message": "string"
}
```

## Link to Issue

Link to this issue: [[FEAT]: API ENDPOINT TO ASSIGN USER'S ROLE IN AN ORGANISATION](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/246)

### CODE REVIEW

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { CreateNewsletterSubscriptionDto } from './dto/create-newsletter-subscri
 import { skipAuth } from '../../helpers/skipAuth';
 import { ApiOperation, ApiParam, ApiResponse, ApiTags } from '@nestjs/swagger';
 import { SuperAdminGuard } from '../../guards/super-admin.guard';
+import { NewsletterSubscriptionResponseDto } from './dto/newsletter-subscription.response.dto';
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:7]: Please why are we reviewing NewsLetter code in adding roles to users

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts [hunk 2]
```diff
@@ -23,9 +24,26 @@ export class NewsletterSubscriptionController {
...
+        data: {
+          type: 'array',
+          items: { $ref: '#/components/schemas/NewsletterSubscriptionResponseDto' },
+        },
+      },
+    },
+  })
+  async findAll(): Promise<{ message: string; data: NewsletterSubscriptionResponseDto[] }> {
+    const subscribers = await this.newsletterSubscriptionService.findAll();
+    return {
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:43]: Use a better descriptive name. 


```suggestion
  async findAllSubscribers(): Promise<{ message: string; data: NewsletterSubscriptionResponseDto[] }> {
    const subscribers = await this.newsletterSubscriptionService.findAllSubscribers();
    return {
```

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts [hunk 3]
```diff
@@ -43,9 +61,27 @@ export class NewsletterSubscriptionController {
   @Get('deleted')
   @HttpCode(HttpStatus.OK)
   @ApiOperation({ summary: 'Fetch all deleted subscribers' })
-  @ApiResponse({ status: 200, type: [Object] })
-  findSoftDeleted() {
-    return this.newsletterSubscriptionService.findSoftDeleted();
+  @ApiResponse({
+    status: 200,
+    description: 'Return all team members',
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:66]:
```suggestion
    <!-- description: 'Return all team members', -->
   description: 'Return all deleted subscribers' 
```

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts [hunk 4]
```diff
@@ -43,9 +61,27 @@ export class NewsletterSubscriptionController {
...
+    schema: {
+      properties: {
+        status: { type: 'string' },
+        message: { type: 'string' },
+        data: {
+          type: 'array',
+          items: { $ref: '#/components/schemas/NewsletterSubscriptionResponseDto' },
+        },
+      },
+    },
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:76]: Please review this schema, there should not be a reason to return all these data when `deleting`.

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts [hunk 5]
```diff
@@ -43,9 +61,27 @@ export class NewsletterSubscriptionController {
...
+  })
+  async findSoftDeleted(): Promise<{ message: string; data: NewsletterSubscriptionResponseDto[] }> {
+    const deletedSubscribers = await this.newsletterSubscriptionService.findSoftDeleted();
+
+    return {
+      message: 'Deleted subscribers list fetched successfully',
+      data: deletedSubscribers,
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:83]: This is not supposed to be returned in a Delete method

File: src/modules/newsletter-subscription/newsletter-subscription.service.ts [hunk 6]
```diff
@@ -24,13 +25,10 @@ export class NewsletterSubscriptionService {
     return response;
   }
 
-  async findAll() {
+  async findAll(): Promise<NewsletterSubscriptionResponseDto[]> {
```
Comment (CFCIfe) [inline src/modules/newsletter-subscription/newsletter-subscription.service.ts:28]:
```suggestion
  async findAllSubscribers(): Promise<NewsletterSubscriptionResponseDto[]> {
```

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts (hunk 2 above)
Comment (CFCIfe) [reply src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:43]: Returns should be done in the Service File

File: src/modules/newsletter-subscription/newsletter-subscription.service.ts (hunk 6 above)
Comment (CFCIfe) [reply src/modules/newsletter-subscription/newsletter-subscription.service.ts:28]: Please implement pagination.

File: src/modules/organisation-role/entities/role.entity.ts [hunk 7]
```diff
@@ -1,11 +0,0 @@
-import { Entity, Column } from 'typeorm';
```
Comment (CFCIfe) [inline src/modules/organisation-role/entities/role.entity.ts:1]: any reason deleting the file?

File: src/modules/organisations/organisations.controller.ts [hunk 8]
```diff
@@ -84,4 +88,47 @@ export class OrganisationsController {
     const { sub } = req.user;
     return this.organisationsService.getOrganisationMembers(org_id, page, page_size, sub);
   }
+
+  @ApiOperation({ summary: 'Get members of an Organisation' })
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.controller.ts:92]: The description of the operation is not correct

File: src/modules/organisations/organisations.controller.ts [hunk 9]
```diff
@@ -84,4 +88,47 @@ export class OrganisationsController {
...
+    status: 403,
+    description: 'User not a member of the organisation',
+  })
+  @Put(':orgId/members/:memberId/role')
+  async updateMemberRole(
+    @Param('memberId') memberId: string,
+    @Param('orgId') orgId: string,
+    @Body() updateMemberRoleDto: UpdateMemberRoleDto
+  ) {
+    try {
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.controller.ts:125]: Remove try...catch.

We catch all `internalServerError` exception and advise not to `throw`

File: src/modules/organisation-role/entities/role.entity.ts (hunk 7 above)
Comment (CFCIfe) [reply src/modules/organisation-role/entities/role.entity.ts:1]: nvm. I see there is another entity file

File: src/modules/newsletter-subscription/newsletter-subscripion.controller.ts (hunk 1 above)
Comment (davidolanrewaju) [reply src/modules/newsletter-subscription/newsletter-subscripion.controller.ts:7]: I created the branch from the branch I used for implementing newsletter subscription

**Pull request closed without merging**